third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp

```diff
@@ -512,18 +492,57 @@ struct TritonAMDGPUInsertInstructionSchedHints
     : public triton::impl::TritonAMDGPUInsertInstructionSchedHintsBase<
           TritonAMDGPUInsertInstructionSchedHints> {
 
+  explicit TritonAMDGPUInsertInstructionSchedHints(StringRef variant) {
+    this->variant = std::move(variant.str());
+  }
+
+  void guardFlashAttentionLikeProblems(triton::FuncOp funcOp) {
+    llvm::SetVector<scf::ForOp> innermostForOps =
+        triton::AMD::getAllInnerForOps(funcOp);
+    for (auto forOp : innermostForOps) {
+      size_t gemmCounter = 0;
+      size_t reduceCounter = 0;
+      bool hasExpOp = false;
+      forOp->walk([&](triton::DotOp) { ++gemmCounter; });
+      forOp->walk([&](triton::ReduceOp) { ++reduceCounter; });
+      forOp->walk([&](math::Exp2Op) { hasExpOp = true; });
+      if ((gemmCounter > 1) && (reduceCounter > 1) && hasExpOp) {
+        OpBuilder builder(forOp->getContext());
+        Block *block = forOp.getBody();
+        builder.setInsertionPoint(block, std::prev(block->end()));
+        builder.create<triton::amdgpu::InstructionSchedHint>(
+            forOp->getLoc(), triton::amdgpu::SchedHint::guard);
+      }
+    }
+  }
+
   void runOnOperation() override {
     MLIRContext *ctx = &getContext();
     ModuleOp mod = getOperation();
 
-    mod.walk([this, ctx](scf::ForOp forOp) {
+    auto maybeSchedHint = triton::amdgpu::symbolizeSchedHint(this->variant);
+    if (!maybeSchedHint) {
+      LDBG("Skipping instruction scheduling: unknown scheduling hint.");
+      return;
+    }
+
+    mod.walk([this](triton::FuncOp funcOp) {
+      guardFlashAttentionLikeProblems(funcOp);
+    });
+
+    triton::amdgpu::SchedHint schedHint = maybeSchedHint.value();
+    if (schedHint == triton::amdgpu::SchedHint::none)
```

Comment:
If we are explicitly told to do nothing (for whatever experiments) we still run the guard at L529? Is that on purpose?